libbeat/outputs/elasticsearch/client_proxy_test.go

```diff
@@ -0,0 +1,246 @@
+	_, err := execClient(
+		"TEST_SERVER_URL="+listeners.server.URL,
+		"TEST_PROXY_URL="+listeners.proxy.URL)
+	assert.NoError(t, err)
+	// We expect one proxy request and 0 server requests
+	listeners.checkFinalState(t, 0, 1)
+}
+
+// TestEnvironmentProxy confirms that we can control the proxy of a client by
+// setting the HTTP_PROXY environment variable (see
+// https://golang.org/pkg/net/http/#ProxyFromEnvironment).
+func TestEnvironmentProxy(t *testing.T) {
+	listeners := testListeners{}
+	listeners.start(t)
+	defer listeners.close()
+
+	// Start a client with HTTP_PROXY set to the proxy listener.
+	// The server is set to a nonexistent URL because ProxyFromEnvironment
+	// always returns a nil proxy for local destination URLs. For this case, we
+	// confirm the intended destination by setting it in the http headers,
+	// triggered in doClientPing by the TEST_HEADER_URL environment variable.
+	_, err := execClient(
+		"TEST_SERVER_URL=http://fakeurl.fake.not-real",
+		"TEST_HEADER_URL="+listeners.server.URL,
+		"HTTP_PROXY="+listeners.proxy.URL)
+	assert.NoError(t, err)
+	// We expect one proxy request and 0 server requests
+	listeners.checkFinalState(t, 0, 1)
+}
+
+// TestClientSettingsOverrideEnvironmentProxy confirms that when both
+// ClientSettings.Proxy and HTTP_PROXY are set, ClientSettings takes precedence.
+func TestClientSettingsOverrideEnvironmentProxy(t *testing.T) {
+	listeners := testListeners{}
+	listeners.start(t)
+	defer listeners.close()
+
+	// Start a client with ClientSettings.Proxy set to the proxy listener and
+	// HTTP_PROXY set to the server listener. We expect that the former will
+	// override the latter and thus we will only see a ping to the proxy.
+	// As above, the fake URL is needed to ensure ProxyFromEnvironment gives a
+	// non-nil result.
+	_, err := execClient(
+		"TEST_SERVER_URL=http://fakeurl.fake.not-real",
+		"TEST_HEADER_URL="+listeners.server.URL,
+		"TEST_PROXY_URL="+listeners.proxy.URL,
+		"HTTP_PROXY="+listeners.server.URL)
+	assert.NoError(t, err)
+	// We expect one proxy request and 0 server requests
+	listeners.checkFinalState(t, 0, 1)
+}
+
+// runClientTest executes the current test binary as a child process,
+// running only the TestClientPing, and calling it with the environment variable
+// TEST_START_CLIENT=1 (so the test can recognize that it is the child process),
+// and any additional environment settings specified in env.
+// This is helpful for testing proxy settings, since we need to have both a
+// proxy / server-side listener and a client that communicates with the server
+// using various proxy settings.
+func execClient(env ...string) (*bytes.Buffer, error) {
+	// The child process always runs only the TestClientPing test, which pings
+	// the server at TEST_SERVER_URL and then terminates.
+	cmd := exec.Command(os.Args[0], "-test.run=TestClientPing")
+	cmd.Env = append(append(os.Environ(),
+		"TEST_START_CLIENT=1"),
+		env...)
+	stderr := new(bytes.Buffer)
```

Comment:
Why are you returning the contents of stderr, if it's ignored when `execClient` is called? Are you using it in a code you have not submitted yet?

Reply:
No, I used this to troubleshoot during development but it's unneeded now; removed.